### PR TITLE
retrieve with redirect (look up in the sameAs of all enabled entities)

### PR DESCRIPTION
**Pull request by SrdjanStevanetic**

### CODE REVIEW

File: entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java [hunk 1]
```diff
@@ -161,6 +162,18 @@ public Optional<EntityRecord> findEntityDupplicationByCoreference(
     EntityRecord value = query.first();
     return Optional.ofNullable(value);
   }
+  
+  public Optional<EntityRecord> findEntityDupplicationBySameAs(String entityId) {
```
Comment (gsergiu) [inline entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java]: The ticket specs indicates that all data.europeana.eu URLs available in same as must be used for searching the entity to redirect to.

File: entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java [hunk 2]
```diff
@@ -161,6 +162,18 @@ public Optional<EntityRecord> findEntityDupplicationByCoreference(
...
+    List<String> listIds = Collections.singletonList(entityId);
+    Query<EntityRecord> query =
+        getDataStore()
+            .find(EntityRecord.class)
+            .disableValidation()
+            .filter(in(ENTITY_SAME_AS, listIds));
+    query.filter(eq(DISABLED, null));
+    EntityRecord value = query.first();
```
Comment (gsergiu) [inline entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java]: we should check if the request returns multiple entities, and probably throw an exception. if we see such cases, we probably have a data issue that needs to be manually corrected in zoho

Comment (gsergiu) [reply entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java]: (note the Zoho deprecation mechanism is not preventing deprecation in loops )

File: entity-management-tests/src/integration-test/java/eu/europeana/entitymanagement/testutils/IntegrationTestUtils.java [hunk 3]
```diff
@@ -247,20 +248,24 @@ public class IntegrationTestUtils {
...
-          AGENT_STALIN_URI, AGENT_STALIN_XML,
-          PLACE_PARIS_URI, PLACE_PARIS_XML,
-          PLACE_HAGENBACH_URI, PLACE_HAGENBACH_XML,
-          TIMESPAN_1ST_CENTURY_URI, TIMESPAN_1ST_CENTURY_XML,
-          CONCEPT_BATHTUB_URI, CONCEPT_BATHTUB_XML,
-          AGENT_JAN_VERMEER_VIAF_URI, AGENT_JAN_VERMEER_XML_VIAF,
-          AGENT_JAN_VERMEER_WIKIDATA_URI, AGENT_JAN_VERMEER_XML_WIKIDATA,
-          AGENT_SCHEGK_URI, AGENT_SCHEGK_XML,
-          AGENT_BIRCH_URI, AGENT_BIRCH_XML);
+  public static final Map<String, String> METIS_RESPONSE_MAP = new HashMap<>(){
```
Comment (gsergiu) [inline entity-management-tests/src/integration-test/java/eu/europeana/entitymanagement/testutils/IntegrationTestUtils.java]: why is this change needed?

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java [hunk 4]
```diff
@@ -851,24 +849,17 @@ private ResponseEntity<String> checkExistingEntity(
...
-              .findFirst();
-      if (sameAsLink.isPresent()) {
-        // return 301 redirect
-        return ResponseEntity.status(HttpStatus.MOVED_PERMANENTLY)
-            .location(URI.create(sameAsLink.get()))
-            .build();
-      }
+    //check if there is any enabled entity in the db, containing the entityId of this entity
+    //in the sameAs ref links and redirect to it
+    Optional<EntityRecord> entityRedirect = entityRecordService.findEntityDupplicationBySameAs(deprecatedEntity.getEntityId());
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java]: the deprecatedEntity.getEntity().getSameReferenceLinks() needs to be used to cover the Concepts as well.
All data.europeana.eu URIs needs to be used for

**Review by gsergiu — CHANGES_REQUESTED** [review]

File: entity-management-tests/src/integration-test/java/eu/europeana/entitymanagement/testutils/IntegrationTestUtils.java (hunk 3 above)
Comment (SrdjanStevanetic) [reply entity-management-tests/src/integration-test/java/eu/europeana/entitymanagement/testutils/IntegrationTestUtils.java]: Because the Map.of is limited to only 10 entries, please see here: https://typeofnan.dev/how-to-create-a-map-of-more-than-10-in-java/. It can also be written the way I did it, but I changed it as it is in the article, with Map.ofEntries.

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java [hunk 5]
```diff
@@ -721,8 +723,31 @@ private ResponseEntity<String> createResponseRetrieve(
...
       throws EuropeanaApiException {
-    EntityRecord entityRecord =
-        entityRecordService.retrieveEntityRecord(type, identifier.toLowerCase(), true);
+    
+    EntityRecord entityRecord = null;
+    try {
+      entityRecord = 
+          entityRecordService.retrieveEntityRecord(type, identifier.toLowerCase(), true);
+    }
+    catch (EntityNotFoundException ex) {
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java]: this is part of the business logic, must be moved to a service method. Something like findRedirection

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java [hunk 6]
```diff
@@ -721,8 +723,31 @@ private ResponseEntity<String> createResponseRetrieve(
...
+        throw new MultipleChoicesException(String.format(MULTIPLE_CHOICES_FOR_REDIRECTION_MSG, entityUri, allCorefUris));
+      }
+      else if(corefEntities.size()==1) {
+        //return 301 redirect
+        return ResponseEntity.status(HttpStatus.MOVED_PERMANENTLY)
+            .location(URI.create(corefEntities.get(0).getEntityId()))
+            .build();
+      }
+      else {
+        throw new EntityNotFoundException(entityUri);
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java]: why not rethrowing the original EntityNotFound exception?

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java [hunk 7]
```diff
@@ -721,8 +723,31 @@ private ResponseEntity<String> createResponseRetrieve(
...
+      }
+    }
+    
     if (entityRecord.isDisabled()) {
       return redirectDeprecated(entityRecord, type, identifier.toLowerCase());
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java]: this method should be refactored as well within this ticket, The service should implement the business logic to select the redirection URI and the controller should only build the response entity

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java [hunk 8]
```diff
@@ -850,25 +875,26 @@ private ResponseEntity<String> checkExistingEntity(
 
   private ResponseEntity<String> redirectDeprecated(
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/EMController.java]: should be refactored to have only one redirection that serves multiple situations (old europeana ID, deprecation, etc.). The service must identify the redirection URL.

**Review by gsergiu — CHANGES_REQUESTED** [review]
